backend/elasticsearch/src/main/java/org/hibernate/search/backend/elasticsearch/search/predicate/impl/ElasticsearchSearchPredicate.java

```diff
@@ -16,17 +16,21 @@
 
 import com.google.gson.JsonObject;
 
-public interface ElasticsearchSearchPredicate extends SearchPredicate {
+public abstract class ElasticsearchSearchPredicate implements SearchPredicate {
 
-	Log log = LoggerFactory.make( Log.class, MethodHandles.lookup() );
+	private static final Log log = LoggerFactory.make( Log.class, MethodHandles.lookup() );
 
-	Set<String> indexNames();
+	public abstract Set<String> indexNames();
 
-	void checkNestableWithin(String expectedParentNestedPath);
+	public abstract void checkNestableWithin(String expectedParentNestedPath);
 
-	JsonObject toJsonQuery(PredicateRequestContext context);
+	public final JsonObject toJsonQuery(PredicateRequestContext context) {
+		return buildJsonQuery( context.predicateContext( this ) );
+	}
+
+	public abstract JsonObject buildJsonQuery(PredicateRequestContext context);
```

Comment:
hmm 😃. My main goal here with this change was to make sure that the context is always created for us and we do not need to think when to create a new one...

> Pushed a bit further, if you called checkNestableWithin whenever a predicate is added to e.g. a boolean predicate clause, this could allow you to throw errors earlier

Wouldn't there be a problem if someone does something like:
```java
var p1 = f.bool().should(f.knn()..); // all looks fine and should would accept the knn
var p2 = f.bool().must(f.match()...).filter(p1); // this should now fail as it would result in a knn predicate being in the filter. But that would mean that we'd need to check p1 is "nestable" and then recursively if any predicates that are nested within p1 are nestable within this new context...
```

I've added a couple of tests for knn in sort/aggregation filters and they are failing (not allow to add knn) right now since Elasticsearch knn predicate just throws exceptions on any nonnull path in `checkNestableWithin` and these filters want to be applied to a field of a nested object.

But then .. I'm probably missing something here 😃

Reply:
> Wouldn't there be a problem if someone does something like:
> 
> ```java
> var p1 = f.bool().should(f.knn()..); // all looks fine and should would accept the knn
> var p2 = f.bool().must(f.match()...).filter(p1); // this should now fail as it would result in a knn predicate being in the filter. But that would mean that we'd need to check p1 is "nestable" and then recursively if any predicates that are nested within p1 are nestable within this new context...
> ```

Why would this be a problem? `ElasticsearchBooleanPredicate` does indeed [check "nestability" recursively](https://github.com/hibernate/hibernate-search/blob/b19562cea319cf90524448f413e4d019b519fd59/backend/elasticsearch/src/main/java/org/hibernate/search/backend/elasticsearch/search/predicate/impl/ElasticsearchBooleanPredicate.java#L66-L72). It always has.

As I mentioned there would be an impact on performance, but I think it's acceptable provided the leaf checks are quick enough.


> I've added a couple of tests for knn in sort/aggregation filters and they are failing (not allow to add knn) right now since Elasticsearch knn predicate just throws exceptions on any nonnull path in checkNestableWithin and these filters want to be applied to a field of a nested object.

These sort/aggregations filters are specifically about nested fields, and nested fields are not supported for knn. So we expect those tests to fail. So all is good, I'm not sure I see the problem?

Reply:
> These sort/aggregations filters are specifically about nested fields, and nested fields are not supported for knn. So we expect those tests to fail. So all is good, I'm not sure I see the problem?

No, there's no problem with that. I was just saying that these cases happened to work as you'd expected them to work purely by chance 😃 

I've removed the `checkNestableWithin` from the building json query and made some more changes around it in the last commit, let me know if that works for you 😃